### ISSUE TITLE
Shader compile error: cannot construct `mat3' from a matrix in GLSL 1.10 (GLSL 1.20 or GLSL ES 1.00 required)

### ISSUE THREAD

**Issue by Echelon9** [scp-fs2open/fs2open.github.com#235]
OpenGL shader compile error reported below:
```
Compiling new shader:
	Model Rendering
   Loading built-in default shader for: main-v.sdr
   Loading built-in default shader for: main-f.sdr
Vertex shader failed to compile:
0:120(44): error: cannot construct `mat3' from a matrix in GLSL 1.10 (GLSL 1.20 or GLSL ES 1.00 required)
0:120(26): error: operands to arithmetic operators must be numeric
0:120(26): error: operands to arithmetic operators must be numeric
0:120(16): error: no matching function for call to `normalize(error)'; candidates are:
0:120(16): error:    float normalize(float)
0:120(16): error:    vec2 normalize(vec2)
0:120(16): error:    vec3 normalize(vec3)
0:120(16): error:    vec4 normalize(vec4)

ERROR! Unable to create vertex shader!
```

Whilst the hardware (Intel Ironlake) isn't state of the art, it should be able to compile or handle downgrade gracefully. Running with `Mesa 10.5.2` on Ubuntu 15.04. Still present with latest Mesa development `Mesa 10.7.0-devel (git-b4b4406)`

```
  OpenGL Vendor    : Intel Open Source Technology Center
  OpenGL Renderer  : Mesa DRI Intel(R) Ironlake Mobile 
  OpenGL Version   : 2.1 Mesa 10.5.2

  Using extension "GL_EXT_fog_coord".
  Using extension "GL_ARB_multitexture".
  Using extension "GL_ARB_texture_env_add".
  Using extension "GL_ARB_texture_compression".
  Using extension "GL_EXT_texture_compression_s3tc".
  Using extension "GL_EXT_texture_filter_anisotropic".
  Using extension "GL_ARB_texture_env_combine".
  Using extension "GL_EXT_compiled_vertex_array".
  Using extension "GL_EXT_draw_range_elements".
  Using extension "GL_ARB_texture_mirrored_repeat".
  Using extension "GL_ARB_texture_non_power_of_two".
  Using extension "GL_ARB_vertex_buffer_object".
  Using extension "GL_ARB_pixel_buffer_object".
  Using extension "GL_SGIS_generate_mipmap".
  Using extension "GL_EXT_framebuffer_object".
  Using extension "GL_ARB_texture_rectangle".
  Using extension "GL_EXT_bgra".
  Using extension "GL_ARB_texture_cube_map".
  Using extension "GL_EXT_texture_lod_bias".
  Using extension "GL_ARB_point_sprite".
  Using extension "GL_ARB_shading_language_100".
  Using extension "GL_ARB_shader_objects".
  Using extension "GL_ARB_vertex_shader".
  Using extension "GL_ARB_fragment_shader".
  Using extension "GL_ARB_shader_texture_lod".
  Using extension "GL_ARB_texture_float".
  Using extension "GL_ARB_draw_elements_base_vertex".
  Using extension "GL_EXT_framebuffer_blit".
  Unable to find extension "GL_EXT_geometry_shader4".
  Using extension "GL_EXT_texture_array".
  Unable to find extension "GL_ARB_uniform_buffer_object".
  Unable to find extension "GL_EXT_transform_feedback".
  Using extension "GL_ARB_draw_instanced".
  Unable to find extension "GL_ARB_texture_buffer_object".
  Found special extension function "glXSwapIntervalSGI".
...
  OpenGL Shader Version: 1.20
```

**Comment by Echelon9** [comment]
Refer similar bug report from SpringRTS and their fix: https://springrts.com/mantis/view.php?id=4643

**Comment by The-E** [comment]
If Mesa is actively lying about their GLSL support, why should we introduce workarounds?

**Comment by Echelon9** [comment]
I don't think it is Mesa lying about their GLSL support though.
My hardware and Mesa version do support through GLSL 1.20, *subject* to the shader actually requesting it.

On non-Apple platforms, we don't add the text `#version 120` to the top of any of our shaders.

Applying the below patch resolved this shader compiler error for me, without having do any hacky intermediate mat3 defines like SpringRTS did. Again, pointing to our application not properly asking for GLSL version of at least 1.20.

```
$ git diff
diff --git a/code/graphics/gropenglshader.cpp b/code/graphics/gropenglshader.cpp
index 875ab14..41eecf6 100644
--- a/code/graphics/gropenglshader.cpp
+++ b/code/graphics/gropenglshader.cpp
@@ -309,9 +309,7 @@ static char *opengl_load_shader(shader_type type_id, char *filename, int flags)
 {
        SCP_string sflags;
 
-#ifdef __APPLE__
-    sflags += "#version 120\n";
-#endif
+       sflags += "#version 120\n";
     
        if (Use_GLSL >= 4) {
                sflags += "#define SHADER_MODEL 4\n";
```

This is going to turn on what the spec says, versus what actual drivers do. But I wouldn't be surprised if the binary blob drivers on Windows and Linux just upscale the GLSL version, whereas Mesa is correctly doing exactly what it is told, though YMMV.

Getting this right could materially improve shader support on Linux for users with Intel hardware, based on what I'm seeing.

**Comment by The-E** [comment]
Congratulations, you just broke shader compilation on recent AMD drivers :P

There's a reason why that particular line is behind a compile-time flag; We only know that it works for sure on the Apple drivers. A fix should therefore be to turn this into a runtime check based on the GL Vendor string, but a blanket solution does not work.

Vertex shader failed to compile:
Vertex shader failed to compile with the following errors:
WARNING: 0:28: warning(#62) enable/warn/disable extension is not found. Extension "GL_ARB_texture_buffer_object" is not supported
ERROR: 0:30: error(#201) Requires extension support: samplerBuffer (GL_EXT_gpu_shader4)
ERROR: 0:70: error(#202) No matching overloaded function found: texelFetch
ERROR: 0:70: error(#160) Cannot convert from: "const float" to: "highp 4-component vector of vec4"
ERROR: 0:71: error(#202) No matching overloaded function found: texelFetch
ERROR: 0:71: error(#160) Cannot convert from: "const float" to: "highp 4-component vector of vec4"
ERROR: 0:72: error(#202) No matching overloaded function found: texelFetch
ERROR: 0:72: error(#160) Cannot convert from: "const float" to: "highp 4-component vector of vec4"
ERROR: 0:73: error(#202) No matching overloaded function found: texelFetch
ERROR: 0:73: error(#160) Cannot convert from: "const float" to: "highp 4-component vector of vec4"
ERROR: error(#273) 9 compilation errors.  No code generated


ERROR! Unable to create vertex shader!

**Comment by Echelon9** [comment]
Wahee! Yes, a proper fix will have this behind some sort of runtime check.
But subject to input from Swifty et al, looks like this is the right track to resolve shaders on Intel Mesa.

**Comment by Echelon9** [comment]
Mesa appears to be behaving exactly per the OpenGL Shading Language specification (v4.40, rev 9):

> Shaders should declare the version of the language they are written to. The language
> version a shader is written to is specified by
>
>  `#version number profileopt`
>
> where number must be a version of the language, following the same convention
> as \_\_VERSION\_\_ above. The directive “#version 440” is required in any shader
> that uses version 4.40 of the language. Any number representing a version of the
> language a compiler does not support will cause a compile-time error to be generated.
>
> **Version 1.10 of the language does not require shaders to include this directive, and
> shaders that do not include a #version directive will be treated as targeting version 1.10.** 


https://www.opengl.org/registry/doc/GLSLangSpec.4.40.pdf (page 17)

**Comment by The-E** [comment]
The problem is that they seem to be alone in implementing the spec as written. Given that we have to support non-Mesa systems, we have to be bug-compatible there.

(And yes, this is a fucked up thing to have to do)

**Comment by asarium** [comment]
Fixed in #316 but requires special handling of the OpenGL implementations @The-E mentioned.

**Comment by Echelon9** [comment]
Just noting here I've been looking at our GLSL shader language interface recently.
It's identified some refactoring opportunities, including the rather horrible `Use_GLSL` variable.
As part of that refactor, I've identified an alternative fix for the GLSL shader conformance issue identified in this issue.

I'm reluctant to draw tester focus away from the soon to be released latest mainline RC, but might get some code out there to a few targeted testers who can provide checks for the key combinations of Linux (proprietary and open source drivers) and Apple across each of Intel, NVIDIA, and AMD Radeon GPUs.

**Comment by Echelon9** [comment]
A short update:
* Mesa-based drivers still report the shader compile error, as our engine doesn't hand its shaders to the drivers' GLSL compiler with a `#version` directive.
* Refactoring of our GLSL setup path and removal of DirectX Shader Model concepts have merged.

A few attempts ( #316, #425 ) have been explored to handle the `#version` standards compliance issue. However so far each has run into issues with the binary blob drivers failing to compile shaders or not dynamically supporting higher feature levels.

**Comment by asarium** [comment]
This should be fixed by #844.
